tests/integration/test_upgrades.py

```diff
@@ -0,0 +1,88 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Superset charm upgrades integration tests."""
+
+import asyncio
+import logging
+
+import pytest
+import pytest_asyncio
+import requests
+from helpers import (
+    APP_NAME,
+    METADATA,
+    POSTGRES_NAME,
+    REDIS_NAME,
+    SUPERSET_SECRET_KEY,
+    get_unit_url,
+    perform_superset_integrations,
+)
+from pytest_operator.plugin import OpsTest
+
+logger = logging.getLogger(__name__)
+
+
+@pytest.mark.skip_if_deployed
+@pytest_asyncio.fixture(name="deploy", scope="module")
+async def deploy(ops_test: OpsTest):
+    """Deploy the app."""
+    asyncio.gather(
+        ops_test.model.deploy(POSTGRES_NAME, channel="14", trust=True),
+        ops_test.model.deploy(REDIS_NAME, channel="edge", trust=True),
+    )
+
+    async with ops_test.fast_forward():
+        await ops_test.model.wait_for_idle(
+            apps=[POSTGRES_NAME, REDIS_NAME],
+            status="active",
+            raise_on_blocked=False,
+            timeout=2000,
+        )
+        superset_config = {
+            "superset-secret-key": SUPERSET_SECRET_KEY,
+        }
+        await ops_test.model.deploy(
+            APP_NAME, channel="stable", config=superset_config
+        )
+        await perform_superset_integrations(ops_test, APP_NAME)
+
+
+@pytest.mark.abort_on_fail
+@pytest.mark.usefixtures("deploy")
+class TestUpgrade:
+    """Integration test for Superset charm upgrade from previous release."""
+
+    async def test_upgrade(self, ops_test: OpsTest):
+        """Builds the current charm and refreshes the current deployment."""
+        charm = await ops_test.build_charm(".")
+        resources = {
+            "superset-image": METADATA["resources"]["superset-image"][
+                "upstream-source"
+            ]
+        }
+
+        await ops_test.model.applications[APP_NAME].refresh(
+            path=str(charm), resources=resources
+        )
+        await ops_test.model.wait_for_idle(
+            apps=[APP_NAME],
+            status="active",
+            raise_on_blocked=False,
+            timeout=600,
+        )
+
+        assert (
+            ops_test.model.applications[APP_NAME].units[0].workload_status
+            == "active"
+        )
+
+    async def test_ui_relation(self, ops_test: OpsTest):
```

Comment:
how confident are we that if this passes, the upgrade was fine?

Reply:
So I was thinking about this when writing it, I could validate the existence of some charts via the UI but in order for those charts to exist I need to `load-examples` as part of the superset charm configuration. When I refresh the charm, had  there been some issue with the upgrade this would not be a reliable data point as the charts would just load again due to the configuration value. As this doesn't add anything additional my reasoning was that this test in combination with the `test_charm_crash` in `test_charm.py` which show that the superset charm itself is stateless and that any changes made are persisted when superset crashes and is re-deployed would be sufficient.

Reply:
Let me see if I can add some validation from the `juju status` channel showing that the channel has been updated